platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ScopedFingerprintWriter.kt

```diff
@@ -21,38 +21,78 @@ import org.gradle.internal.serialize.graph.CloseableWriteContext
 import org.gradle.internal.serialize.graph.runWriteOperation
 import org.gradle.internal.serialize.graph.withPropertyTrace
 import java.io.Closeable
+import java.util.ArrayDeque
+import java.util.Queue
 
 
-internal
-class ScopedFingerprintWriter<T>(
+internal class ScopedFingerprintWriter<T>(
     private val writeContext: CloseableWriteContext
 ) : Closeable {
+    private class PendingWrite<out T>(val value: T, val trace: PropertyTrace?)
+
+    private val pendingWrites: Queue<PendingWrite<T>> = ArrayDeque()
+
+    private var isWriting = false
+
     override fun close() {
         // we synchronize access to all resources used by callbacks
         // in case there was still an event being dispatched at closing time.
         synchronized(writeContext) {
-            unsafeWrite(null)
+            require(!isWriting && pendingWrites.isEmpty()) {
+                "Incomplete write detected"
+            }
+
+            writeValueLocked(null)
             writeContext.close()
         }
     }
 
     fun write(value: T, trace: PropertyTrace? = null) {
         synchronized(writeContext) {
-            withPropertyTrace(trace) {
-                unsafeWrite(value)
+            if (isWriting) {
+                // Writing a value may cause another value to be written, e.g. when `provider {}` used as a ValueSource parameter reads an environment variable.
+                // These are configuration inputs, but we cannot write them immediately, or we will break the layout of the outer value.
+                // It doesn't expect that an Input comes right in a middle of a Provider's serialized form.
```

Comment:
These comments are helpful, but I can't follow this specific sentence. Maybe rephrase it?

Reply:
> These comments are helpful, but I can't follow this specific sentence. Maybe rephrase it?

Thanks, I've elaborated the description.